File: benchmarks/correctness/real_world/r/translate.R
library(Biostrings)
library(jsonlite)
seqs <- readDNAStringSet("real_data/yeast_genome.fa")
names(seqs) <- sub(" .*$", "", names(seqs))
results <- list()
for (i in 1:min(3, length(seqs))) {
  sub_seq <- subseq(seqs[[i]], 1, min(99, nchar(seqs[[i]])))
  dna_str <- as.character(sub_seq)
  protein <- as.character(translate(sub_seq))
  # BioLang's translate() ends at the first stop codon; Biostrings emits '*'
  # and continues. Split on a literal '*' — Biostrings masks base::strsplit.
  protein <- unlist(base::strsplit(protein, "*", fixed = TRUE))[1]
  if (is.na(protein)) protein <- ""
  id <- sub("\\s.*", "", names(seqs)[i])
  results[[i]] <- list(id = id, dna = toupper(dna_str), protein = protein)
}
cat(toJSON(list(translations = results), auto_unbox = TRUE, digits = 10))
cat("\n")

File: benchmarks/correctness/real_world/python/translate.py
import json
from Bio import SeqIO
from Bio.Seq import Seq

results = []
for i, record in enumerate(SeqIO.parse("real_data/yeast_genome.fa", "fasta")):
    if i >= 3:
        break
    sub = str(record.seq)[:99].upper()
    # to_stop matches BioLang's translate(), which ends at the first stop
    # codon rather than emitting '*' and continuing.
    protein = str(Seq(sub).translate(to_stop=True))
    results.append({"id": record.id, "dna": sub, "protein": protein})

print(json.dumps({"translations": results}))
